### DISCUSSION TITLE
Firmware build, Make error (2)

### DISCUSSION THREAD

**Discussion by WvBijsterveld** [danielnilsson9/bbs-fw#180] [Q&A] [answered]
Tried to build the HEX firmware file, my make was misfunctioning alot, ended up getting that all in order. (sort of)
Then ran into a error which i have no explanation for
![afbeelding](https://github.com/danielnilsson9/bbs-fw/assets/14361841/bd9dda35-aacd-465e-b7c3-2f343e45cb96)

Was used on 1.4.0 main branch and EigenMania's pretension patch.

`PS E:\Bafang-config\BBSFW-Source-build\bbs-fw-feature-apply_pretension\src\firmware> mingw32-make.exe TARGET_CONTROLLER=BBSHD
Building bbs-fw for BBSHD
sdcc -o throttle.rel -c -I./ -Ibbsx -Ddouble=float --std-c99 -DBBSHD -mmcs51 --model-large --xram-size 3840      throttle.c
process_begin: CreateProcess(NULL, sdcc -o throttle.rel -c -I./ -Ibbsx -Ddouble=float --std-c99 -DBBSHD -mmcs51 --model-large --xram-size 3840 throttle.c, ...) failed.
make (e=2): The system cannot find the file specified.
makefile:50: recipe for target 'throttle.rel' failed
mingw32-make: *** [throttle.rel] Error 2
PS E:\Bafang-config\BBSFW-Source-build\bbs-fw-feature-apply_pretension\src\firmware>`

Any help would be much appreciated!

Also for anyone that is affraid to try yourself or doesn't know anything about building etc. after this is fixed i wanna make a discussion post where i post my "how-to" of how to build, and more(i think)

**Comment by EigenMania** [comment]
I would guess that Windows cannot find sdcc. Check that it's correctly added to your PATH.

**Reply by WvBijsterveld** [reply]
it was shown when viewing the folder / sln in vs studio or code in the build interface, i installed it on 2 machines and both ticked the box where it said add to path. on main pc i checked the path for make and that’s where i also saw SDCC, could be wrong so will check again.

also if i don’t make changes to makefile the controller target error comes up no matter what i do.

Thank you for your reply and again any help would be greatly appreciated

**Reply by EigenMania** [reply]
Do you get the same error when trying to build any of the other .c files with sdcc? For example have you tried just executing:

sdcc -o app.rel -c -I./ -Ibbsx -Ddouble=float --std-c99 -DBBSHD -mmcs51 --model-large --xram-size 3840	 app.c

or the same for any of the other files? You should be able to run this alone on the command line (on PS, Terminal, cmd, ... whatever you use).

**Reply by WvBijsterveld** [reply]
pasted your command into PS opened in firmware source folder and got this.

`app.h:57: warning 283: function declarator with no prototype
app.h:59: warning 283: function declarator with no prototype
app.h:66: warning 283: function declarator with no prototype
app.h:67: warning 283: function declarator with no prototype
app.h:68: warning 283: function declarator with no prototype
cfgstore.h:138: warning 283: function declarator with no prototype
cfgstore.h:140: warning 283: function declarator with no prototype
cfgstore.h:141: warning 283: function declarator with no prototype
cfgstore.h:143: warning 283: function declarator with no prototype
cfgstore.h:144: warning 283: function declarator with no prototype
motor.h:19: warning 283: function declarator with no prototype
motor.h:22: warning 283: function declarator with no prototype
motor.h:24: warning 283: function declarator with no prototype
motor.h:25: warning 283: function declarator with no prototype
motor.h:27: warning 283: function declarator with no prototype
motor.h:28: warning 283: function declarator with no prototype
motor.h:29: warning 283: function declarator with no prototype
motor.h:36: warning 283: function declarator with no prototype
motor.h:37: warning 283: function declarator with no prototype
motor.h:38: warning 283: function declarator with no prototype
sensors.h:17: warning 283: function declarator with no prototype
sensors.h:18: warning 283: function declarator with no prototype
sensors.h:21: warning 283: function declarator with no prototype
sensors.h:22: warning 283: function declarator with no prototype
sensors.h:23: warning 283: function declarator with no prototype
sensors.h:24: warning 283: function declarator with no prototype
sensors.h:27: warning 283: function declarator with no prototype
sensors.h:28: warning 283: function declarator with no prototype
sensors.h:30: warning 283: function declarator with no prototype
sensors.h:31: warning 283: function declarator with no prototype
sensors.h:33: warning 283: function declarator with no prototype
sensors.h:34: warning 283: function declarator with no prototype
sensors.h:36: warning 283: function declarator with no prototype
sensors.h:37: warning 283: function declarator with no prototype
throttle.h:18: warning 283: function declarator with no prototype
throttle.h:19: warning 283: function declarator with no prototype
lights.h:16: warning 283: function declarator with no prototype
lights.h:18: warning 283: function declarator with no prototype
lights.h:19: warning 283: function declarator with no prototype
uart.h:15: warning 283: function declarator with no prototype
uart.h:17: warning 283: function declarator with no prototype
uart.h:18: warning 283: function declarator with no prototype
uart.h:21: warning 283: function declarator with no prototype
eventlog.h:73: warning 283: function declarator with no prototype
system.h:22: warning 283: function declarator with no prototype
system.h:24: warning 283: function declarator with no prototype
app.c:82: warning 283: function declarator with no prototype
app.c:85: warning 283: function declarator with no prototype
app.c:89: warning 283: function declarator with no prototype
app.c:125: warning 283: function declarator with no prototype
app.c:277: warning 283: function declarator with no prototype
app.c:282: warning 283: function declarator with no prototype
app.c:338: warning 283: function declarator with no prototype
app.c:896: warning 283: function declarator with no prototype
app.c:919: warning 283: function declarator with no prototype`

**Reply by EigenMania** [reply]
But it looks like it ran? Can you confirm if app.rel was built in the firmware folder? The warnings are not important / critical. Just the compiler complaining that the function declarations in the headers don't have any information about potential inputs.

**Reply by WvBijsterveld** [reply]
will check in a moment, is SDCC used to make instead of using make in itself?
Will report back when i have the chance to check the files.

**Reply by WvBijsterveld** [reply]
i do have a file named app.rel and lst .sym

Now what do i do for the firmware file?

**Reply by WvBijsterveld** [reply]
i ran command 

sdcc -o main.hex -c -I./ -Ibbsx -Ddouble=float --std-c99 -DBBSHD -mmcs51 --model-large --xram-size 3840 main.c

Got a error that there was no intellisense in bbx so i pasted the one on fw to fw/bbx then file was created but it doesn't look like it's actually "it" compared to the prebuild hex file that is available.
[main.txt](https://github.com/danielnilsson9/bbs-fw/files/12522079/main.txt)
renamed from .hex to .txt



Tried something else;

code ran :    make -B TARGET_CONTROLLER=BBSHD -k

got errors again for files missing i.e. stc15.h while located in bbx

i ran with NO -k first try but after setting -k i saw all fatal errors
It's all missing file, listing all files that are supposedly missing
bbsx/stc15.h (2x)
eeprom.h
lights.h
motor.h
sensors.h
system.h
uart.h
watchdog.h

**Reply by EigenMania** [reply]
![image](https://github.com/danielnilsson9/bbs-fw/assets/23363130/92753619-2521-47c8-aa64-1479c608991c)

Here is the (successful) build output from VS. This should give you some idea of what commands are run (and in what order) for a successful build. The final line is a bit long and truncated in the picture: 

sdcc -o bbs-fw.ihx -I./ -Ibbsx -Ddouble=float --std-c99 -DBBSHD -mmcs51 --model-large --xram-size 3840	 main.c app.rel battery.rel cfgstore.rel eventlog.rel extcom.rel throttle.rel bbsx/adc.rel bbsx/eeprom.rel bbsx/lights.rel bbsx/motor.rel bbsx/sensors.rel bbsx/system.rel bbsx/timers.rel bbsx/uart.rel bbsx/watchdog.rel

**Reply by WvBijsterveld** [reply]
I tried running all code using PS, runs good until the first bbsx command, the adc.c one, throws errors file not found, also i got this on my pc but now sometimes on my laptop aswell (gave up on pc lol)  

warning 190: ISO C forbids an empty translation unit

Don't know if this affects it badly or is just a regular error.(found something online about fixing it but didn't work and the error didn't show on laptop so didn't bother anymore)

I really wanna try to make this work,  @EigenMania  could you help me with this(the building from source)?

Also can you compile the bbs-fw-pretension patch and send it to me please i want to build files but all this time i spent into it and i don't even have the firmware on my controller is meh :/

Back to the problem solving; I opened the  .sln  file in vs however if i try to build it immediatly throws the error of target controller thing failed, i think it's supposed to be click and build so i don't know why it's doing that.

What file do you open in VS, exactly what build options and whatnot do you use? any changes to any files or configs or PREF in VS/extensions, oh yea any essential extensions(i have a few that was recommend in vs studio and through the internet.

Thank you in advance!

**Reply by EigenMania** [reply]
[BBSFWTool.exe.txt](https://github.com/danielnilsson9/bbs-fw/files/12534110/BBSFWTool.exe.txt)
[bbs-fw.hex.txt](https://github.com/danielnilsson9/bbs-fw/files/12534116/bbs-fw.hex.txt)

Here it is, looking forward for your feedback. 

The apply_pretension feature is branched off of master, so has everything else from the 1.4.0 release. For me building in VS works out of the box, with no changes to the makefile.

**Reply by WvBijsterveld** [reply]
Thanks!

The FW tool did build however with nothing done to it(atleast that's what i remember could be something minor)
So don't know what's holding back the FW.

Will flash in the afternoon and will report any feedback in a few days time.

**Comment by WvBijsterveld** [comment, marked as answer]
LISTEN YOU DO NOT KNOW HOW HAPPY AND DISAPPOINTED I AM AT THE SAME TIME

After spending days just endlessy trying bullshit hours on end and when not trying something it just sitting in the back of my mind.

I am not experienced at all (changed values in config files of games like money) Started a course Software development just a week ago but IT WAS JUST MY FUCKING HEAD.

Yabber yabber above solution below!

When clicking the link for installing SDCC, it takes you to sourceforge to install the LATEST release, HOWEVER instead i saw it but never thought twice about the 4.2 version **needed**(to date it auto downloads 4.3.

In short, have make functioning (check that mingw/bin/ is added to path, alot of instructions of adding make to work)
Download the SDCC **### _4.2_** here: [https://sourceforge.net/projects/sdcc/files/sdcc-win64/4.2.0/sdcc-4.2.0-x64-setup.exe/download](https://sourceforge.net/projects/sdcc/files/sdcc-win64/4.2.0/sdcc-4.2.0-x64-setup.exe/download)
@danielnilsson9 Can you change the link to the link provided above? thank you in advance!

Shift right click when in /src/firmware/ folder when installing the source zip and extracting it somewhere. then click open in terminal/cmd/powershell.
Then run make -b TARGET_CONTROLLER=(your controller here BBSHD,BBS02 or TSDZ2)                          Without ()



Will post a tutorial very soon (mention me please if i go inactive/don't do it) to never let anyone make the same mistake as me, aswell as to document my progress and will probably try to explain how to change CONSTANTS like max temp for the people that want that or are curious about that.

And thank you @EigenMania for helping and responding because who knows maybe i would have gave up with no reply from anyone.

Will try to contribute more and more to this project from now on.

**Reply by WvBijsterveld** [reply]
TLDR: Installed SDCC 4.3 instead of 4.2 REQUIRED.

The rest is just dumb words. and me trying to explain how to build like i haven't just spend all my hours in this so tiny mistake (i tried 2 machines (one was a clean win11 build)
